src/modules/base.ts

```diff
@@ -1,119 +1,138 @@
 /**
  * Base module for all API modules
  *
- * @module modules/base
- *
  * @since 1.0.0
  */
 import commsUtils from '../utils/comms';
 import {name as PROTOCOL_AGENT_NAME, version as PROTOCOL_AGENT_VERSION} from '../../package.json';
 
 const PROTOCOL_NAME = 'purecloud-client-apps';
 
-const ACTION_NAME_KEY = 'action';
-const PROTOCOL_NAME_KEY = 'protocol';
-const PROTOCOL_AGENT_NAME_KEY = 'protocolAgentName';
-const PROTOCOL_AGENT_VERSION_KEY = 'protocolAgentVersion';
+interface ProtocolDetails {
+    name: string;
+    agentName: string;
+    agentVersion: string;
+}
+
+export interface SDKMessagePayload {
+    action: string;
+    hook: string;
+    protocol: string;
+    protocolAgentName: string;
+    protocolAgentVersion: string;
+}
+
+export type MessageListener = (event: SDKMessagePayload) => void;
+export type MessagePayloadFilter = (payload: SDKMessagePayload) => boolean;
+
+interface ListenerOptions {
+    /** 
+     * Set to true to only invoke this listener once; it will be removed after first invocation.
+     * false by default.  null/undefined will use default.
+     */
+    once: boolean;
```

Comment:
The comment mentions it being optional, but it's not marked as optional (but`msgPayloadFilter` is).   Looks like it's usually used like `Partial<ListenerOptions>`, but maybe it'd be better to do `type ListenerOptions = Partial<{/*...*/}>` instead.

Reply:
Yep, I like this better

Reply:
Yeah, this works really nice.  I think the existing js woud have technically allowed null for once but no point in exposing this flexibility to the typed users.  undefined or boolean works perfectly well here.  The normalize will also protect us from js users, so this looks great to me.